### ISSUE TITLE
Running prediction on images

### ISSUE THREAD

**Issue by Jamaslab** [SciSharp/TensorFlow.NET#400]
Hi all,
I created the following model
```
public class DnCNN {
        private const int batch_size = 128;

        Tensor X, Y_, Y;
        Tensor loss;
        Operation optimizer;

        Session sess;

        public DnCNN() {
            X = tf.placeholder(tf.float32, shape: (-1, -1, -1, 3), name: "input_image");
            Y_ = tf.placeholder(tf.float32, shape: (-1, -1, -1, 3), name: "clean_image");

            Y = BuildModel(X);

            loss = (1.0 / batch_size) * tf.nn.relu(Y_ - Y);
            optimizer = tf.train.AdamOptimizer(0.001f, name: "AdamOptimizer").minimize(loss);

            sess = new Session();
            var init = tf.global_variables_initializer();
            sess.run(init);
        }

        private Tensor BuildModel(Tensor input, bool is_training = true) {
            var output = tf.layers.conv2d(input, 64, new int[] { 3, 3 }, name: "conv1", padding: "same");
            for (int i = 2; i < 20; i++) {
                output = tf.layers.conv2d(output, 64, new int[] { 3, 3 }, name: "conv" + i, padding: "same", use_bias: false);
            }
            output = tf.layers.conv2d(output, 3, new int[] { 3, 3 }, name: "conv20", padding: "same", use_bias: false);

            return input - output;
        }

        public void Train(List<Bitmap> inputImages, List<Bitmap> outputImages, int epochs = 1) {
            var sw = new Stopwatch();
            sw.Start();
            NDArray x_train = GenerateDataset(inputImages);
            NDArray y_train = GenerateDataset(outputImages);

            var saver = new Saver();
            print($"Dataset created in {sw.ElapsedMilliseconds}ms");
            sw.Restart();
            for (int i = 0; i < epochs; i++) {
                sess.run(optimizer, (X, new Tensor(x_train)), (Y_, new Tensor(y_train)));
                // Calculate and display the batch loss and accuracy
                var result = sess.run(new[] { loss }, new FeedItem(X, x_train), new FeedItem(Y_, y_train));
                print($"iter {i.ToString("000")}: {sw.ElapsedMilliseconds}ms");
                sw.Restart();

                saver.save(sess, @"E:\Downloads\ciao.ckpt");
            }
            
        }

        public Bitmap Evaluate(List<Bitmap> inputImages) {
            var sw = new Stopwatch();
            sw.Start();
            NDArray x_eval= GenerateDataset(inputImages);
            var output = sess.run(Y, new FeedItem(X, x_eval));
            sw.Stop();
            print($"Inference done in {sw.ElapsedMilliseconds}ms");

            output = 255*output;
            return image(output[0].astype(NPTypeCode.Byte));
     }
}
```

However, when I'm trying to evaluate the model, i get an overflow error in the conversion of the output to Byte. If I check the values of the output, I get also negative values. I guess i'm doing something wrong or during the training or during the evaluation. Could you help me out?

**Comment by Oceania2018** [comment]
@Jamaslab Can you attach the working python version here? It will be helpful.

**Comment by Jamaslab** [comment]
Sorry, I'm developing this directly in C#. Truly I should develop this in python first.

**Comment by Nucs** [comment]
Please provide us with a zip of 4-5 images for input and output so we can reproduce locally. Please verify that it occurs when using these images.

**Comment by Jamaslab** [comment]
Attached the dataset. You can use the same image list for input and output and execution for the moment.

[github_dataset.zip](https://github.com/SciSharp/TensorFlow.NET/files/3657082/github_dataset.zip)

**Comment by Nucs** [comment]
It doesn't seem to be a problem with the library [heres](https://gist.github.com/Nucs/4cd1220cc676945447f663ee93114578) the code I ran locally that uses [NumSharp.Bitmap](https://www.nuget.org/packages/NumSharp.Bitmap/)

**Comment by Jamaslab** [comment]
Hi, 
I still get the same problem when running the evaluate function. Your code was missing the conversion to float32 for the training and the normalization to [0,1].

When I run evaluate on one of the same picture, I get the following error at the line
`return image(output[0].astype(NPTypeCode.Byte));`

OverflowException: Overflow_Byte

**Comment by Nucs** [comment]
Ok so I added normalization to [0, 1]
You get `Overflow_Byte` because its possible to get a value higher than 1.0 with CNN when no activation is specified.
From Evaluate method:
![image](https://user-images.githubusercontent.com/649919/65871391-f52cfd80-e386-11e9-8fa6-09fefb428566.png)
So when call:
```C#
output = 255 * output;
return image(output[0].astype(NPTypeCode.Byte));
```
Your values get above 255 and converting to byte throws overflow.

**Comment by Jamaslab** [comment]
Ok, thanks for the suggestion. I modified the model to have it like this now:
```
private Tensor BuildModel(Tensor input) {
    var output = tf.layers.conv2d(input, 64, new int[] { 3, 3 }, name: "conv1", padding: "same", activation: tf.nn.relu());
    for (int i = 2; i < 20; i++) {
         output = tf.layers.conv2d(output, 64, new int[] { 3, 3 }, name: "conv" + i, padding: "same", use_bias: false);
        output = tf.nn.relu(tf.layers.batch_normalization(output, training: is_training));
    }
    output = tf.layers.conv2d(output, 3, new int[] { 3, 3 }, name: "conv20", padding: "same", use_bias: false);

    return input - output;
}
```
as it is done here https://github.com/wbhu/DnCNN-tensorflow/blob/master/model.py.
However now the tf.layers.batch_normalization throws a null reference exception.

**Comment by Nucs** [comment]
Take a look at #404, It is likely that the exception is getting swallowed. Take a look at the console for printed exceptions.
I'll take a deeper look later.

**Comment by Jamaslab** [comment]
Not sure this may be the cause but in BatchNormalization.cs, in the function call:
       ```
 protected override Tensor call(Tensor inputs, Tensor training = null)
        {
            Tensor outputs = null;

            if (fused)
            {
                outputs = _fused_batch_norm(inputs, training: training);
                return outputs;
            }

            throw new NotImplementedException("BatchNormalization call");
        }
```

The tensor training is null, even if I'm passing a tensor for the training during the definition. Later on, the following function is called:
`var results = tf_utils.smart_cond(training, _fused_batch_norm_training, _fused_batch_norm_inference);`

but as training tensor is null, the method fails.

**Comment by Oceania2018** [comment]
@Jamaslab Is the [python code](https://github.com/wbhu/DnCNN-tensorflow/blob/master/model.py) what you want to run in tf.net?

**Comment by Jamaslab** [comment]
Alright, my bad (**very bad**). I was initializing the training tensor after.
Now the error appears here:
 ```
public void _add_control_input(Operation op)
        {
            //c_api.TF_AddControlInput(_operDesc, op);
            c_api.AddControlInput(graph, _handle, op);
        }
```

`System.EntryPointNotFoundException: 'Unable to find an entry point named 'AddControlInput' in DLL 'tensorflow'.'`

@Oceania2018 I'm building the model as it is built here:
https://github.com/wbhu/DnCNN-tensorflow/blob/master/model.py

**Comment by Oceania2018** [comment]
This issue will be fixed until Tensorflow v1.15 released. You have to wait one more week I think.
